src/main/java/org/reaktivity/nukleus/mqtt/internal/stream/MqttServerFactory.java

```diff
@@ -1102,18 +1164,34 @@ private void onDecodeUnsubscribe(
             final int packetId = unsubscribe.packetId();
 
             MqttTopicFW topic;
+            boolean error = false;
             for (int progress = offset; progress < limit; progress = topic.limit())
             {
                 topic = mqttTopicRO.tryWrap(buffer, progress, limit);
                 if (topic == null || topic.filter() == null)
                 {
-                    onDecodeError(traceId, authorization, PROTOCOL_ERROR);
-                    return;
+                    error = true;
+                    break;
                 }
-                final int topicKey = topicKey(topic.filter().asString());
-                subscribers.remove(topicKey);
+                final String topicString = topic.filter().asString();
+                final int topicKey = topicKey(topicString);
+                final MqttServerStream stream = streams.get(topicKey);
+
+                final MqttEndExFW endEx = mqttEndExRW.wrap(extBuffer, 0, extBuffer.capacity())
+                                              .typeId(mqttTypeId)
+                                              .build();
+
+                stream.doApplicationEnd(traceId, authorization, endEx);
```

Comment:
What does the spec say about partially processing partially invalid `UNSUBSCRIBE` packets?
Is it all-or-nothing? If so then you'll need to do two passes over the topic filters to verify they are all valid before processing any of them.
Though if there is a decode error, then I suppose we are going to clean up everything anyway, so probably it is a moot point, agreed?

Reply:
Yeah, agreed. Either way should honestly work the same.